web/packages/teleport/src/Integrations/Enroll/Shared/state.ts

```diff
@@ -0,0 +1,213 @@
+  useEffect,
+  useRef,
+  useState,
+  type Dispatch,
+  type SetStateAction,
+} from 'react';
+import { useHistory } from 'react-router';
+
+import { isIntegrationTag, type IntegrationTag } from './common';
+
+const sortKeys = ['name'] as const;
+
+export type IntegrationPickerSortKey = (typeof sortKeys)[number] | undefined;
+export type IntegrationPickerSortDirection = 'ASC' | 'DESC';
+
+function isValidSortKey(key: string): key is IntegrationPickerSortKey {
+  return sortKeys.includes(key as IntegrationPickerSortKey);
+}
+
+export interface IntegrationPickerFilters {
+  tags: IntegrationTag[];
+}
+
+export interface IntegrationPickerState {
+  filters: IntegrationPickerFilters;
+  search: string;
+  sortKey: IntegrationPickerSortKey;
+  sortDirection: IntegrationPickerSortDirection;
+}
+
+export type IntegrationPickerFilterKey = keyof IntegrationPickerFilters;
+
+export function searchParamsToState(params: URLSearchParams) {
+  const state: IntegrationPickerState = {
+    filters: {
+      tags: [],
+    },
+    search: '',
+    sortKey: undefined,
+    sortDirection: 'ASC',
+  };
+
+  const sortKey = params.get('sort');
+  if (sortKey && isValidSortKey(sortKey)) {
+    state.sortKey = sortKey as IntegrationPickerSortKey;
+  }
+
+  const direction = params.get('direction');
+  if (direction === 'ASC' || direction === 'DESC') {
+    state.sortDirection = direction as IntegrationPickerSortDirection;
+  }
+
+  const tags = params.getAll('tags');
+  if (tags.length > 0) {
+    state.filters.tags = tags.filter(isIntegrationTag);
+  }
+
+  const search = params.get('search');
+  if (search) {
+    state.search = search;
+  }
+
+  return state;
+}
+
+export function stateToSearchParams(state: IntegrationPickerState) {
+  const urlParams = new URLSearchParams();
+
+  if (state.filters.tags.length > 0) {
+    for (const tags of state.filters.tags) {
+      urlParams.append('tags', tags);
+    }
+  }
+
+  if (state.search) {
+    urlParams.set('search', state.search);
+  }
+
+  if (state.sortKey !== undefined) {
+    urlParams.set('direction', state.sortDirection);
+    urlParams.set('sort', state.sortKey);
+  }
+
+  return urlParams.toString();
+}
+
+function sortArray(array: string[]) {
+  return array.toSorted((a, b) => a.localeCompare(b));
+}
+
+function arraysAreEqual(a: string[], b: string[]) {
+  if (a.length !== b.length) {
+    return false;
+  }
+
+  return JSON.stringify(sortArray(a)) === JSON.stringify(sortArray(b));
+}
+
+function filtersAreEqual(
+  a: IntegrationPickerFilters,
+  b: IntegrationPickerFilters
+) {
+  if (a.tags.length !== b.tags.length) {
+    return false;
+  }
+
+  return arraysAreEqual(a.tags, b.tags);
+}
+
+export function statesAreEqual(
+  a: IntegrationPickerState,
+  b: IntegrationPickerState
+) {
+  return (
+    a.sortKey === b.sortKey &&
+    a.sortDirection === b.sortDirection &&
+    filtersAreEqual(a.filters, b.filters) &&
+    a.search === b.search
+  );
+}
+
+export function useIntegrationPickerState(): [
```

Comment:
Callout: This is mostly repurposed from https://github.com/gravitational/teleport/blob/master/web/packages/teleport/src/SessionRecordings/list/state.ts but this hook feels like it could be generic 🤔

Reply:
It might be a good idea to try this if it’s not too much work. But I'd do this outside this PR, since it's already quite large.